# sourmash tutorials and notebooks

## The first two tutorials!

These tutorials are both command line tutorials that should work on Mac OS
X and Linux. They require about 5 GB of disk space and 5 GB of RAM.

* [The first sourmash tutorial - making signatures, comparing, and searching](tutorial-basic.html)

* [Using sourmash LCA to do taxonomic classification](tutorials-lca.html)

## Background and details

These next three tutorials are all notebooks that you can view, run
yourself, or run interactively online via the
[binder](http://mybinder.org) service.

* [An introduction to k-mers for genome comparison and analysis](kmers-and-minhash.html)

* [Some sourmash command line examples!](sourmash-examples.html)

* [Working with private collections of signatures.](sourmash-collections.html)

## More information

If you are a Python programmer, you might also be interested in our [API examples](api-example.html) as well as a short guide to [Using the `LCA_Database` API.](using-LCA-database-API.html)

If you prefer R, we have [a short guide to using sourmash output with R](other-languages.html).
